File: backend/src/main/java/com/keycloak/saas/config/ApplicationConfigService.java
package com.keycloak.saas.config;

import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.cache.annotation.CacheEvict;
import org.springframework.cache.annotation.Cacheable;
import org.springframework.dao.DataIntegrityViolationException;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.Instant;
import java.time.OffsetDateTime;
import java.time.ZoneOffset;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.stream.Collectors;

/**
 * Service for managing application-wide configuration.
 * All config changes are logged to audit_logs for compliance tracking.
 * In-memory cache (TTL configurable) reduces DB queries.
 */
@Slf4j
@Service
@RequiredArgsConstructor
public class ApplicationConfigService {

    private final ApplicationConfigRepository configRepository;
    private final AuditLogService auditLogService;
    private final LoggerService loggerService;

    private static final String CACHE_NAME = "application_config";

    // ========================================================================
    // GET OPERATIONS
    // ========================================================================

    /**
     * Get config value by key (cached).
     * @param configKey The configuration key
     * @return Config value as String, or empty Optional if not found
     */
    @Cacheable(value = CACHE_NAME, key = "#configKey")
    public Optional<String> getConfigValue(String configKey) {
        loggerService.debug("Fetching config value for key: {}", configKey);
        Optional<ApplicationConfigEntity> config = configRepository.findByConfigKeyAndIsActiveTrue(configKey);
        if (config.isPresent()) {
            loggerService.debug("Config found for key: {} = {}", configKey,
                    config.get().isSecret() ? "***MASKED***" : config.get().getConfigValue());
            return Optional.of(config.get().getConfigValue());
        }
        loggerService.warn("Config key not found: {}", configKey);
        return Optional.empty();
    }

    /**
     * Get config value with default fallback.
     * @param configKey The configuration key
     * @param defaultValue Value to return if key not found
     * @return Config value or default
     */
    public String getConfigValueOrDefault(String configKey, String defaultValue) {
        loggerService.info("Fetching config with default fallback. Key: {}, Default: {}",
                configKey, defaultValue);
        return getConfigValue(configKey).orElse(defaultValue);
    }

    /**
     * Get config as typed value (Boolean, Integer, etc).
     * @param configKey The configuration key
     * @param type Target type class
     * @return Typed value
     */
    public <T> T getConfigAs(String configKey, Class<T> type) {
        loggerService.debug("Fetching config as type: {} for key: {}", type.getSimpleName(), configKey);
        Optional<String> value = getConfigValue(configKey);
        if (value.isEmpty()) {
            loggerService.warn("Config key not found for type conversion: {}", configKey);
            return null;
        }
        return convertValue(value.get(), type);
    }

    /**
     * Get all active configuration entries.
     * @return Map of key -> value pairs (active configs only)
     */
    public Map<String, String> getAllConfig() {
        loggerService.info("Fetching all active configuration entries");
        List<ApplicationConfigEntity> configs = configRepository.findAllByIsActiveTrue();
        loggerService.debug("Found {} active config entries", configs.size());
        return configs.stream()
                .collect(Collectors.toMap(
                        ApplicationConfigEntity::getConfigKey,
                        ApplicationConfigEntity::getConfigValue
                ));
    }

    /**
     * Get config entity by ID (includes audit trail).
     * @param configId The configuration ID
     * @return Config entity with all metadata
     */
    public Optional<ApplicationConfigEntity> getConfigById(Integer configId) {
        loggerService.debug("Fetching config by ID: {}", configId);
        return configRepository.findById(configId);
    }

    // ========================================================================
    // CREATE/UPDATE OPERATIONS (WITH AUDIT LOGGING)
    // ========================================================================

    /**
     * Create new configuration entry.
     * Logs audit event on success.
     * @param configKey Unique configuration key
     * @param configValue Configuration value
     * @param configType Type (STRING, INTEGER, BOOLEAN, etc)
     * @param description Human-readable description
     * @param isSecret Whether value should be masked in logs
     * @param createdBy User creating the config
     * @return Saved configuration entity
     */
    @Transactional
    public ApplicationConfigEntity createConfig(String configKey, String configValue, String configType,
                                               String description, boolean isSecret, String createdBy) {
        loggerService.info("Creating new config entry. Key: {}, Type: {}, Secret: {}",
                configKey, configType, isSecret);

        try {
            ApplicationConfigEntity config = ApplicationConfigEntity.builder()
                    .configKey(configKey)
                    .configValue(configValue)
                    .configType(configType)
                    .description(description)
                    .isSecret(isSecret)
                    .isActive(true)
                    .createdBy(createdBy)
                    .createdAt(OffsetDateTime.now(ZoneOffset.UTC))
                    .version(1)
                    .build();

            ApplicationConfigEntity saved = configRepository.save(config);
            loggerService.info("Config created successfully. ID: {}, Key: {}", saved.getConfigId(), configKey);

            auditLogService.logEvent(AuditEventBuilder.builder()
                    .eventType("CONFIG_CREATED")
                    .eventCategory("CONFIGURATION")
                    .action("CREATE")
                    .resourceType("APPLICATION_CONFIG")
                    .resourceId(saved.getConfigId().toString())
                    .newValue(isSecret ? "***MASKED***" : configValue)
                    .userId(createdBy)
                    .status("SUCCESS")
                    .build());

            invalidateCache(configKey);
            return saved;
        } catch (DataIntegrityViolationException e) {
            loggerService.error("Config key already exists: {}. Error: {}", configKey, e.getMessage());
            auditLogService.logEvent(AuditEventBuilder.builder()
                    .eventType("CONFIG_CREATE_FAILED")
                    .eventCategory("CONFIGURATION")
                    .action("CREATE")
                    .resourceType("APPLICATION_CONFIG")
                    .resourceId(configKey)
                    .status("FAILURE")
                    .errorMessage("Duplicate config key: " + configKey)
                    .userId(createdBy)
                    .build());
            throw new IllegalArgumentException("Config key already exists: " + configKey);
        } catch (Exception e) {
            loggerService.error("Failed to create config: {}", configKey, e);
            throw e;
        }
    }

    /**
     * Update existing configuration entry.
     * Logs old and new values for audit trail.
     * @param configId Configuration ID to update
     * @param configValue New value
     * @param updatedBy User performing update
     * @return Updated configuration entity
     */
    @Transactional
    public ApplicationConfigEntity updateConfig(Integer configId, String configValue, String updatedBy) {
        loggerService.info("Updating config. ID: {}, UpdatedBy: {}", configId, updatedBy);

        ApplicationConfigEntity config = configRepository.findById(configId)
                .orElseThrow(() -> {
                    loggerService.error("Config not found for update. ID: {}", configId);
                    return new IllegalArgumentException("Config not found: " + configId);
                });

        String oldValue = config.getConfigValue();
        config.setConfigValue(configValue);
        config.setUpdatedAt(OffsetDateTime.now(ZoneOffset.UTC));
        config.setUpdatedBy(updatedBy);
        config.setVersion(config.getVersion() + 1);

        ApplicationConfigEntity updated = configRepository.save(config);
        loggerService.info("Config updated successfully. ID: {}, Key: {}, Version: {}",
                configId, config.getConfigKey(), config.getVersion());

        auditLogService.logEvent(AuditEventBuilder.builder()
                .eventType("CONFIG_UPDATED")
                .eventCategory("CONFIGURATION")
                .action("UPDATE")
                .resourceType("APPLICATION_CONFIG")
                .resourceId(configId.toString())
                .oldValue(config.isSecret() ? "***MASKED***" : oldValue)
                .newValue(config.isSecret() ? "***MASKED***" : configValue)
                .userId(updatedBy)
                .status("SUCCESS")
                .build());

        invalidateCache(config.getConfigKey());
        return updated;
    }

    /**
     * Soft-delete configuration (mark as inactive).
     * Hard deletes are not allowed for audit compliance.
     * @param configId Configuration ID to deactivate
     * @param deactivatedBy User performing deactivation
     */
    @Transactional
    public void deactivateConfig(Integer configId, String deactivatedBy) {
        loggerService.info("Deactivating config. ID: {}, DeactivatedBy: {}", configId, deactivatedBy);

        ApplicationConfigEntity config = configRepository.findById(configId)
                .orElseThrow(() -> {
                    loggerService.error("Config not found for deactivation. ID: {}", configId);
                    return new IllegalArgumentException("Config not found: " + configId);
                });

        config.setIsActive(false);
        config.setUpdatedAt(OffsetDateTime.now(ZoneOffset.UTC));
        config.setUpdatedBy(deactivatedBy);

        configRepository.save(config);
        loggerService.info("Config deactivated successfully. ID: {}, Key: {}", configId, config.getConfigKey());

        auditLogService.logEvent(AuditEventBuilder.builder()
                .eventType("CONFIG_DEACTIVATED")
                .eventCategory("CONFIGURATION")
                .action("DEACTIVATE")
                .resourceType("APPLICATION_CONFIG")
                .resourceId(configId.toString())
                .oldValue("is_active=true")
                .newValue("is_active=false")
                .userId(deactivatedBy)
                .status("SUCCESS")
                .build());

        invalidateCache(config.getConfigKey());
    }

    // ========================================================================
    // HELPER METHODS
    // ========================================================================

    /**
     * Invalidate cache for specific config key.
     * @param configKey The configuration key
     */
    @CacheEvict(value = CACHE_NAME, key = "#configKey")
    public void invalidateCache(String configKey) {
        loggerService.debug("Cache invalidated for key: {}", configKey);
    }

    /**
     * Invalidate all application config cache.
     */
    @CacheEvict(value = CACHE_NAME, allEntries = true)
    public void invalidateAllCache() {
        loggerService.debug("All application config cache invalidated");
    }

    /**
     * Convert string value to target type.
     * @param value String value to convert
     * @param type Target type class
     * @return Converted value
     */
    @SuppressWarnings("unchecked")
    private <T> T convertValue(String value, Class<T> type) {
        loggerService.debug("Converting value '{}' to type: {}", value, type.getSimpleName());
        try {
            if (type == String.class) {
                return (T) value;
            } else if (type == Boolean.class || type == boolean.class) {
                return (T) Boolean.valueOf(value);
            } else if (type == Integer.class || type == int.class) {
                return (T) Integer.valueOf(value);
            } else if (type == Long.class || type == long.class) {
                return (T) Long.valueOf(value);
            } else if (type == Double.class || type == double.class) {
                return (T) Double.valueOf(value);
            }
            loggerService.warn("Unsupported type conversion: {}", type.getSimpleName());
            return null;
        } catch (Exception e) {
            loggerService.error("Failed to convert value '{}' to type: {}", value, type.getSimpleName(), e);
            throw new IllegalArgumentException("Cannot convert '" + value + "' to " + type.getSimpleName());
        }
    }
}

File: backend/src/services/ApplicationConfigService.ts
import pino from 'pino';
import { ApplicationConfig, AuditEvent } from '../types/entities.js';

const logger = pino();

export class ApplicationConfigService {
  private cache: Map<string, string> = new Map();
  private cacheExpiry: Map<string, number> = new Map();
  private readonly cacheTTL = 5 * 60 * 1000; // 5 minutes

  constructor(
    private configRepository: any,
    private auditLogService: any,
    private loggerService: any
  ) {}

  async getConfigValue(configKey: string): Promise<string | null> {
    this.loggerService.debug(`Fetching config value for key: ${configKey}`);

    // Check cache first
    const cached = this.cache.get(configKey);
    const expiry = this.cacheExpiry.get(configKey);
    if (cached && expiry && Date.now() < expiry) {
      this.loggerService.debug(`Config found in cache for key: ${configKey}`);
      return cached;
    }

    const config = await this.configRepository.findByKeyAndActive(configKey);
    if (config) {
      this.loggerService.debug(`Config found for key: ${configKey}`);
      this.cache.set(configKey, config.configValue);
      this.cacheExpiry.set(configKey, Date.now() + this.cacheTTL);
      return config.configValue;
    }

    this.loggerService.warn(`Config key not found: ${configKey}`);
    return null;
  }

  async getConfigValueOrDefault(configKey: string, defaultValue: string): Promise<string> {
    this.loggerService.info(`Fetching config with default fallback. Key: ${configKey}, Default: ${defaultValue}`);
    const value = await this.getConfigValue(configKey);
    return value || defaultValue;
  }

  async getConfigAs<T>(configKey: string, type: new (...args: unknown[]) => T): Promise<T | null> {
    this.loggerService.debug(`Fetching config as type: ${type.name} for key: ${configKey}`);
    const value = await this.getConfigValue(configKey);
    if (!value) {
      this.loggerService.warn(`Config key not found for type conversion: ${configKey}`);
      return null;
    }
    return this.convertValue(value, type);
  }

  async getAllConfig(): Promise<Record<string, string>> {
    this.loggerService.info('Fetching all active configuration entries');
    const configs = await this.configRepository.findAllActive();
    this.loggerService.debug(`Found ${configs.length} active config entries`);
    return configs.reduce((acc: Record<string, string>, cfg: ApplicationConfig) => {
      acc[cfg.configKey] = cfg.configValue;
      return acc;
    }, {});
  }

  async getConfigById(configId: number): Promise<ApplicationConfig | null> {
    this.loggerService.debug(`Fetching config by ID: ${configId}`);
    return this.configRepository.findById(configId);
  }

  async createConfig(
    configKey: string,
    configValue: string,
    configType: string,
    description: string,
    isSecret: boolean,
    createdBy: string
  ): Promise<ApplicationConfig> {
    this.loggerService.info(`Creating new config entry. Key: ${configKey}, Type: ${configType}, Secret: ${isSecret}`);

    try {
      const config: Partial<ApplicationConfig> = {
        configKey,
        configValue,
        configType,
        description,
        isSecret,
        isActive: true,
        createdBy,
        createdAt: new Date(),
        version: 1
      };

      const saved = await this.configRepository.save(config);
      this.loggerService.info(`Config created successfully. ID: ${saved.configId}, Key: ${configKey}`);

      const auditEvent: AuditEvent = {
        eventType: 'CONFIG_CREATED',
        eventCategory: 'CONFIGURATION',
        action: 'CREATE',
        resourceType: 'APPLICATION_CONFIG',
        resourceId: saved.configId.toString(),
        newValue: isSecret ? '***MASKED***' : configValue,
        userId: createdBy,
        status: 'SUCCESS',
        logLevel: 'INFO'
      };
      await this.auditLogService.logEvent(auditEvent);

      this.invalidateCache(configKey);
      return saved;
    } catch (e) {
      const error = e as Error;
      if (error.message.includes('duplicate') || error.message.includes('UNIQUE')) {
        this.loggerService.error(`Config key already exists: ${configKey}. Error: ${error.message}`);
        const failureEvent: AuditEvent = {
          eventType: 'CONFIG_CREATE_FAILED',
          eventCategory: 'CONFIGURATION',
          action: 'CREATE',
          resourceType: 'APPLICATION_CONFIG',
          resourceId: configKey,
          status: 'FAILURE',
          errorMessage: `Duplicate config key: ${configKey}`,
          userId: createdBy,
          logLevel: 'ERROR'
        };
        await this.auditLogService.logEvent(failureEvent);
        throw new Error(`Config key already exists: ${configKey}`);
      }
      this.loggerService.error(`Failed to create config: ${configKey}`, error);
      throw e;
    }
  }

  async updateConfig(configId: number, configValue: string, updatedBy: string): Promise<ApplicationConfig> {
    this.loggerService.info(`Updating config. ID: ${configId}, UpdatedBy: ${updatedBy}`);

    const config = await this.configRepository.findById(configId);
    if (!config) {
      this.loggerService.error(`Config not found for update. ID: ${configId}`);
      throw new Error(`Config not found: ${configId}`);
    }

    const oldValue = config.configValue;
    config.configValue = configValue;
    config.updatedAt = new Date();
    config.updatedBy = updatedBy;
    config.version = (config.version || 1) + 1;

    const updated = await this.configRepository.save(config);
    this.loggerService.info(`Config updated successfully. ID: ${configId}, Key: ${config.configKey}, Version: ${config.version}`);

    const auditEvent: AuditEvent = {
      eventType: 'CONFIG_UPDATED',
      eventCategory: 'CONFIGURATION',
      action: 'UPDATE',
      resourceType: 'APPLICATION_CONFIG',
      resourceId: configId.toString(),
      oldValue: config.isSecret ? '***MASKED***' : oldValue,
      newValue: config.isSecret ? '***MASKED***' : configValue,
      userId: updatedBy,
      status: 'SUCCESS',
      logLevel: 'INFO'
    };
    await this.auditLogService.logEvent(auditEvent);

    this.invalidateCache(config.configKey);
    return updated;
  }

  async deactivateConfig(configId: number, deactivatedBy: string): Promise<void> {
    this.loggerService.info(`Deactivating config. ID: ${configId}, DeactivatedBy: ${deactivatedBy}`);

    const config = await this.configRepository.findById(configId);
    if (!config) {
      this.loggerService.error(`Config not found for deactivation. ID: ${configId}`);
      throw new Error(`Config not found: ${configId}`);
    }

    config.isActive = false;
    config.updatedAt = new Date();
    config.updatedBy = deactivatedBy;

    await this.configRepository.save(config);
    this.loggerService.info(`Config deactivated successfully. ID: ${configId}, Key: ${config.configKey}`);

    const auditEvent: AuditEvent = {
      eventType: 'CONFIG_DEACTIVATED',
      eventCategory: 'CONFIGURATION',
      action: 'DEACTIVATE',
      resourceType: 'APPLICATION_CONFIG',
      resourceId: configId.toString(),
      oldValue: 'is_active=true',
      newValue: 'is_active=false',
      userId: deactivatedBy,
      status: 'SUCCESS',
      logLevel: 'INFO'
    };
    await this.auditLogService.logEvent(auditEvent);

    this.invalidateCache(config.configKey);
  }

  private invalidateCache(configKey: string): void {
    this.cache.delete(configKey);
    this.cacheExpiry.delete(configKey);
    this.loggerService.debug(`Cache invalidated for key: ${configKey}`);
  }

  private invalidateAllCache(): void {
    this.cache.clear();
    this.cacheExpiry.clear();
    this.loggerService.debug('All application config cache invalidated');
  }

  private convertValue<T>(value: string, type: new (...args: unknown[]) => T): T {
    this.loggerService.debug(`Converting value '${value}' to type: ${type.name}`);
    try {
      if (type === String) return value as unknown as T;
      if (type === Boolean) return (value.toLowerCase() === 'true') as unknown as T;
      if (type === Number) return (Number(value)) as unknown as T;
      if (type === BigInt) return (BigInt(value)) as unknown as T;
      this.loggerService.warn(`Unsupported type conversion: ${type.name}`);
      return null as unknown as T;
    } catch (e) {
      this.loggerService.error(`Failed to convert value '${value}' to type: ${type.name}`, e as Error);
      throw new Error(`Cannot convert '${value}' to ${type.name}`);
    }
  }
}
